src/Cli/dotnet/commands/dotnet-workload/search/WorkloadSearchVersionsCommand.cs

```diff
@@ -123,5 +125,47 @@ public override int Execute()
 
             return 0;
         }
+
+        private List<string> GetVersions(int numberOfWorkloadSetsToTake)
+        {
+            var featureBand = new SdkFeatureBand(_sdkVersion);
+            var packageId = _installer.GetManifestPackageId(new ManifestId("Microsoft.NET.Workloads"), featureBand);
+
+            try
+            {
+                return PackageDownloader.GetLatestPackageVersions(packageId, numberOfWorkloadSetsToTake, packageSourceLocation: null, includePreview: _includePreviews)
+                    .GetAwaiter().GetResult()
+                    .Select(version => WorkloadSetVersion.FromWorkloadSetPackageVersion(featureBand, version.ToString()))
+                    .ToList();
+            }
+            catch (NuGetPackageNotFoundException)
+            {
+                Microsoft.DotNet.Cli.Utils.Reporter.Error.WriteLine(string.Format(LocalizableStrings.NoWorkloadVersionsFound, featureBand));
+                return null;
+            }
+        }
+
+        public string FindBestWorkloadSetFromComponents()
+        {
+            var versions = GetVersions(0);
+            if (versions is null)
+            {
+                return null;
+            }
+
+            var workloadVersions = _workloadVersion.Split(';');
+            var packageNamesAndVersions = workloadVersions.Select(version =>
+            {
+                var split = version.Split('@');
+                return (new ManifestId(_resolver.GetManifestFromWorkload(new WorkloadId(split[0])).Id), new ManifestVersion(split[1]));
```

Comment:
`GetManifestFromWorkload` has to actually download and extract the NuGet package.  We might want to add some parallelism to those downloads, but it's probably a bit tricky to do so safely, so I would leave it as it is for now.